File: ruby/040.rb
def find_result(candidates, i, target, results, list)
  if target == 0
    results.push(list.dup)
    return
  end

  return if target < 0
  return if i == candidates.length

  i.upto(candidates.length - 1).each do |index|
    next if index > i && candidates[index] == candidates[index - 1]
    list.push(candidates[index])
    find_result(candidates, index + 1, target - candidates[index], results, list)
    list.pop
  end
end

# @param {Integer[]} candidates
# @param {Integer} target
# @return {Integer[][]}
def combination_sum2(candidates, target)
  candidates.sort!
  results = []
  list = []
  find_result(candidates, 0, target, results, list)
  results
end

File: javascript/040.js
/**
 * @param {number[]} candidates
 * @param {number} target
 * @return {number[][]}
 */
var combinationSum2 = function(candidates, target) {
  candidates.sort((a, b) => a - b);

  let results = [];
  let result = [];
  const findResults = (startIndex, target) => {
    if (target === 0) {
      results.push(result.slice(0));
      return;
    }
    if (target < 0) return;
    if (startIndex === candidates.length) return;

    for (let i = startIndex; i < candidates.length; ++i) {
      if (i > startIndex && candidates[i] === candidates[i - 1]) continue;
      result.push(candidates[i]);
      findResults(i + 1, target - candidates[i]);
      result.pop();
    }
  };
  findResults(0, target);

  return results;
};